File: frontend/src/types/collaboration.ts
export type WorkspaceRole = "viewer" | "contributor" | "admin";

export type SharedDataType =
  | "rfp_summary"
  | "compliance_matrix"
  | "proposal_section"
  | "forecast"
  | "contract_feed";

export interface SharedWorkspace {
  id: number;
  owner_id: number;
  rfp_id?: number | null;
  name: string;
  description?: string | null;
  member_count: number;
  created_at: string;
  updated_at: string;
}

export interface WorkspaceInvitation {
  id: number;
  workspace_id: number;
  email: string;
  role: WorkspaceRole;
  accept_token?: string | null;
  is_accepted: boolean;
  expires_at: string;
  created_at: string;
}

export interface WorkspaceMember {
  id: number;
  workspace_id: number;
  user_id: number;
  role: WorkspaceRole;
  user_email?: string | null;
  user_name?: string | null;
  created_at: string;
}

export interface SharedDataPermission {
  id: number;
  workspace_id: number;
  data_type: SharedDataType;
  entity_id: number;
  label?: string | null;
  requires_approval: boolean;
  approval_status: "pending" | "approved" | "revoked";
  approved_by_user_id?: number | null;
  approved_at?: string | null;
  expires_at?: string | null;
  partner_user_id?: number | null;
  created_at: string;
}

export interface ContractFeedCatalogItem {
  id: number;
  name: string;
  source: string;
  description: string;
}

export interface ContractFeedPresetItem {
  key: string;
  name: string;
  description: string;
  feed_ids: number[];
}

export interface SharePresetApplyResponse {
  preset_key: string;
  applied_count: number;
  shared_items: SharedDataPermission[];
}

export interface ShareGovernanceSummary {
  workspace_id: number;
  total_shared_items: number;
  pending_approval_count: number;
  approved_count: number;
  revoked_count: number;
  expired_count: number;
  expiring_7d_count: number;
  scoped_share_count: number;
  global_share_count: number;
}

export interface ShareGovernanceTrendPoint {
  date: string;
  shared_count: number;
  approvals_completed_count: number;
  approved_within_sla_count: number;
  approved_after_sla_count: number;
  average_approval_hours?: number | null;
}

export interface ShareGovernanceTrends {
  workspace_id: number;
  days: number;
  sla_hours: number;
  overdue_pending_count: number;
  sla_approval_rate: number;
  points: ShareGovernanceTrendPoint[];
}

export interface GovernanceAnomaly {
  code: string;
  severity: "info" | "warning" | "critical";
  title: string;
  description: string;
  metric_value: number;
  threshold: number;
  recommendation: string;
}

export interface ComplianceDigestSchedule {
  workspace_id: number;
  user_id: number;
  frequency: "daily" | "weekly";
  day_of_week: number | null;
  hour_utc: number;
  minute_utc: number;
  channel: "in_app" | "email";
  anomalies_only: boolean;
  is_enabled: boolean;
  last_sent_at: string | null;
}

export interface ComplianceDigestPreview {
  workspace_id: number;
  generated_at: string;
  summary: ShareGovernanceSummary;
  trends: ShareGovernanceTrends;
  anomalies: GovernanceAnomaly[];
  schedule: ComplianceDigestSchedule;
}

export interface PortalView {
  workspace_name: string;
  workspace_description?: string | null;
  rfp_title?: string | null;
  shared_items: SharedDataPermission[];
  members: WorkspaceMember[];
}

// Real-Time Presence & Section Locking

export interface DocumentPresenceUser {
  user_id: number;
  user_name: string;
  joined_at: string;
}

export interface SectionLock {
  section_id: number;
  user_id: number;
  user_name: string;
  proposal_id?: number | null;
  locked_at: string;
}

export interface DocumentPresence {
  proposal_id: number;
  users: DocumentPresenceUser[];
  locks: SectionLock[];
}

// Inbox

export type InboxMessageType = "general" | "opportunity_alert" | "rfp_forward";

export interface InboxMessage {
  id: number;
  workspace_id: number;
  sender_id: number;
  sender_name?: string | null;
  sender_email?: string | null;
  subject: string;
  body: string;
  message_type: InboxMessageType;
  is_read: boolean;
  read_by: number[];
  attachments: string[];
  created_at: string;
  updated_at: string;
}

export interface InboxListResponse {
  items: InboxMessage[];
  total: number;
  page: number;
  page_size: number;
}

// Cursor Presence

export interface CursorPosition {
  user_id: number;
  user_name: string;
  section_id: number | null;
  position: number | null;
  color?: string;
  timestamp?: string;
  updated_at?: string;
}

File: backend/app/schemas/collaboration.py
"""
RFP Sniper - Collaboration Schemas
====================================
Request/response models for workspaces, invitations, and data sharing.
"""

from datetime import datetime

from pydantic import BaseModel


class WorkspaceCreate(BaseModel):
    name: str
    rfp_id: int | None = None
    description: str | None = None


class WorkspaceUpdate(BaseModel):
    name: str | None = None
    description: str | None = None


class WorkspaceRead(BaseModel):
    id: int
    owner_id: int
    rfp_id: int | None = None
    name: str
    description: str | None = None
    member_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class InvitationCreate(BaseModel):
    email: str
    role: str = "viewer"


class InvitationRead(BaseModel):
    id: int
    workspace_id: int
    email: str
    role: str
    accept_token: str | None = None
    is_accepted: bool
    expires_at: datetime
    created_at: datetime

    model_config = {"from_attributes": True}


class MemberRead(BaseModel):
    id: int
    workspace_id: int
    user_id: int
    role: str
    user_email: str | None = None
    user_name: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class MemberRoleUpdate(BaseModel):
    role: str


class ShareDataCreate(BaseModel):
    data_type: str  # rfp_summary, compliance_matrix, proposal_section, forecast, contract_feed
    entity_id: int
    requires_approval: bool = False
    expires_at: datetime | None = None
    partner_user_id: int | None = None
    step_up_code: str | None = None


class SharedDataRead(BaseModel):
    id: int
    workspace_id: int
    data_type: str
    entity_id: int
    label: str | None = None
    requires_approval: bool
    approval_status: str
    approved_by_user_id: int | None = None
    approved_at: datetime | None = None
    expires_at: datetime | None = None
    partner_user_id: int | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class ContractFeedCatalogItem(BaseModel):
    id: int
    name: str
    source: str
    description: str


class ContractFeedPresetItem(BaseModel):
    key: str
    name: str
    description: str
    feed_ids: list[int]


class SharePresetCreate(BaseModel):
    preset_key: str
    step_up_code: str | None = None


class SharePresetApplyResponse(BaseModel):
    preset_key: str
    applied_count: int
    shared_items: list[SharedDataRead]


class ShareGovernanceSummaryRead(BaseModel):
    workspace_id: int
    total_shared_items: int
    pending_approval_count: int
    approved_count: int
    revoked_count: int
    expired_count: int
    expiring_7d_count: int
    scoped_share_count: int
    global_share_count: int


class ShareGovernanceTrendPointRead(BaseModel):
    date: str
    shared_count: int
    approvals_completed_count: int
    approved_within_sla_count: int
    approved_after_sla_count: int
    average_approval_hours: float | None = None


class ShareGovernanceTrendRead(BaseModel):
    workspace_id: int
    days: int
    sla_hours: int
    overdue_pending_count: int
    sla_approval_rate: float
    points: list[ShareGovernanceTrendPointRead]


class GovernanceAnomalyRead(BaseModel):
    code: str
    severity: str
    title: str
    description: str
    metric_value: float
    threshold: float
    recommendation: str


class ComplianceDigestScheduleRead(BaseModel):
    workspace_id: int
    user_id: int
    frequency: str
    day_of_week: int | None = None
    hour_utc: int
    minute_utc: int
    channel: str
    recipient_role: str
    anomalies_only: bool
    is_enabled: bool
    last_sent_at: datetime | None = None


class ComplianceDigestScheduleUpdate(BaseModel):
    frequency: str = "weekly"
    day_of_week: int | None = 1
    hour_utc: int = 13
    minute_utc: int = 0
    channel: str = "in_app"
    recipient_role: str = "all"
    anomalies_only: bool = False
    is_enabled: bool = True


class ComplianceDigestDeliveryRead(BaseModel):
    id: int
    workspace_id: int
    user_id: int
    schedule_id: int | None = None
    status: str
    attempt_number: int
    retry_of_delivery_id: int | None = None
    channel: str
    recipient_role: str
    recipient_count: int
    anomalies_count: int
    failure_reason: str | None = None
    generated_at: datetime
    created_at: datetime


class ComplianceDigestDeliverySummaryRead(BaseModel):
    total_attempts: int
    success_count: int
    failed_count: int
    retry_attempt_count: int
    last_status: str | None = None
    last_failure_reason: str | None = None
    last_sent_at: datetime | None = None


class ComplianceDigestDeliveryListRead(BaseModel):
    workspace_id: int
    user_id: int
    summary: ComplianceDigestDeliverySummaryRead
    items: list[ComplianceDigestDeliveryRead]


class ComplianceDigestPreviewRead(BaseModel):
    workspace_id: int
    generated_at: datetime
    recipient_role: str
    recipient_count: int
    summary: ShareGovernanceSummaryRead
    trends: ShareGovernanceTrendRead
    anomalies: list[GovernanceAnomalyRead]
    schedule: ComplianceDigestScheduleRead
    delivery_summary: ComplianceDigestDeliverySummaryRead | None = None


class PortalView(BaseModel):
    """Read-only partner portal view of a workspace."""

    workspace_name: str
    workspace_description: str | None = None
    rfp_title: str | None = None
    shared_items: list[SharedDataRead] = []
    members: list[MemberRead] = []
